compiler/src/main/scala/org/scalajs/core/compiler/GenJSExports.scala

```diff
@@ -1058,45 +1061,79 @@ trait GenJSExports[G <: Global with Singleton] extends SubComponent {
+          mutable = false, rest = false)
+    }
 
-  private def genFormalArgRef(index: Int, minArgc: Int)(
-      implicit pos: Position): js.Tree = {
-    if (index <= minArgc)
-      js.VarRef(js.Ident("arg$" + index))(jstpe.AnyType)
-    else
-      js.JSBracketSelect(genRestArgRef(), js.IntLiteral(index - 1 - minArgc))
-  }
+    def genFormalArgs()(implicit pos: Position): List[js.ParamDef] = {
+      val fixedParamDefs = fixedParamNames.toList.map { paramName =>
+        js.ParamDef(js.Ident(paramName), jstpe.AnyType,
+            mutable = false, rest = false)
+      }
 
-  private def genVarargRef(fixedParamCount: Int, minArgc: Int)(
-      implicit pos: Position): js.Tree = {
-    val restParam = genRestArgRef()
-    assert(fixedParamCount >= minArgc,
-        s"genVarargRef($fixedParamCount, $minArgc) at $pos")
-    if (fixedParamCount == minArgc) restParam
-    else {
-      js.JSBracketMethodApply(restParam, js.StringLiteral("slice"), List(
-          js.IntLiteral(fixedParamCount - minArgc)))
+      if (needsRestParam) {
+        val restParamDef = js.ParamDef(js.Ident(restParamName), jstpe.AnyType,
+            mutable = false, rest = true)
+        fixedParamDefs :+ restParamDef
+      } else {
+        fixedParamDefs
+      }
+    }
+
+    def genArgRef(index: Int)(implicit pos: Position): js.Tree = {
+      if (index < minArgc)
+        js.VarRef(js.Ident(fixedParamNames(index)))(jstpe.AnyType)
+      else
+        js.JSBracketSelect(genRestArgRef(), js.IntLiteral(index - minArgc))
+    }
+
+    def genVarargRef(fixedParamCount: Int)(implicit pos: Position): js.Tree = {
+      val restParam = genRestArgRef()
+      assert(fixedParamCount >= minArgc,
+          s"genVarargRef($fixedParamCount) with minArgc = $minArgc at $pos")
+      if (fixedParamCount == minArgc) {
+        restParam
+      } else {
+        js.JSBracketMethodApply(restParam, js.StringLiteral("slice"), List(
+            js.IntLiteral(fixedParamCount - minArgc)))
+      }
+    }
+
+    def genRestArgRef()(implicit pos: Position): js.Tree = {
+      assert(needsRestParam,
+          s"trying to generate a reference to non-existent rest param at $pos")
+      js.VarRef(js.Ident(restParamName))(jstpe.AnyType)
+    }
+
+    def genAllArgsRefsForForwarder()(implicit pos: Position): List[js.Tree] = {
+      val fixedArgRefs = fixedParamNames.toList.map { paramName =>
+        js.VarRef(js.Ident(paramName))(jstpe.AnyType)
+      }
+
+      if (needsRestParam) {
+        val restArgRef = js.VarRef(js.Ident(restParamName))(jstpe.AnyType)
+        fixedArgRefs :+ restArgRef
+      } else {
+        fixedArgRefs
+      }
     }
   }
 
-  private def genRestArgRef()(implicit pos: Position): js.Tree =
-    js.VarRef(js.Ident("arg$rest"))(jstpe.AnyType)
+  private object FormalArgsRegistry {
+    def forGetter(): FormalArgsRegistry = new FormalArgsRegistry(0, false)
+
+    def forSetter(): FormalArgsRegistry = new FormalArgsRegistry(1, false)
+  }
```

Comment:
It feels to me that these are better inlined at their respective call sites. Like that `FormalArgsRegistry` has no knowledge about setters / getters. This means in practice that destructuring of `genFormalArgs` will have to take place when generating a setter. However, IMHO this is not any less clean than asserting in `genSetterFormalArg`.

Reply:
Makes sense. Updated.